### ISSUE TITLE
ICML writer improperly wraps template variables?

### ISSUE THREAD

**Issue by richarddb** [jgm/pandoc#3551]
I've been working to modify the ICML output template to be suitable for use as a letter template. So in the pandoc markdown file, I prepend a YAML section with variables that would be used as front/end-matter on the letter. For example:

```
---
sender: My Name
salutation: Dear John
close: Sincerely
...
```
I then cloned the default.icml to add the front/end-matter content into the template. For example, I would add the specification for the use of the sender name against a defined paragraph style in the ICML:

```
<ParagraphStyleRange AppliedParagraphStyle="ParagraphStyle/Sender%3aSender Name">
  <CharacterStyleRange AppliedCharacterStyle="CharacterStyle/Default">
    <Content>$sender$</Content>
    <Br />
  </CharacterStyleRange>
</ParagraphStyleRange>
```

The problem is my use of the variable `sender`. When I try substitute the variable, extra content gets added around the variable. So if `$sender$ := "Sender Name"` I would get this:

```
<Content><CharacterStyleRange AppliedCharacterStyle="$ID/NormalCharacterStyle">
	<Content>Sender Name</Content>
</CharacterStyleRange></Content>
```

...instead of just this:

```
<Content>Sender Name</Content>
```

In other words, the `<Content><CharacterStyleRange AppliedCharacterStyle="$ID/NormalCharacterStyle">` and `</CharacterStyleRange></Content>` are getting pre/post-pended to the content I expect. As a workaround, I could remove the character definition from my template; but then I'd be lacking the required line break `<Br />` before the closing `</CharacterStyleRange>`.

I had a look in the [ICML writer](https://github.com/jgm/pandoc/blob/master/src/Text/Pandoc/Writers/ICML.hs). It seems to me that the function on line 520 is what's adding the additional tags. (But then this is the first time I'm reading Haskell code.)

**Comment by jgm** [comment]
The metadata fields are interpreted as Markdown. You can see how they're represented internally by doing `pandoc -s -t native`.  You'll see that `sender` is

    MetaInlines [Str "My",Space,Str "Name"]

So now you can see why you get this extra material.  This is part of how the ICML writer renders a `Str` element.

For your purposes, couldn't you just use a template like

    <Content>$sender$</Content><Br />

so the `<CharacterStyleRange>` element would be inside `<Content>` rather than outside?  Or is that illegal ICML?

Note:  although the pandoc document model allows pure strings to be included in metadata (`MetaString` vs `MetaInlines`), there is no way to force a metadata field to be interpreted as a pure string (or for that matter as "raw" content which wouldn't even be escaped in the output).  There's a relevant issue here: #2139.

**Comment by richarddb** [comment]
This is an interesting problem... I can see that sometimes you'd need one behaviour (with the wrapping), other times not.

Would it make sense to support both? So something like $var$ will output the variable as if it were text in the format; but $var.raw()$ wouldn't? The reason I'm suggesting the latter format, is that you could extend it to other functions... I don't know if it's out of bounds for pandoc, but I'm thinking there's some lightweight templating functionality that would push it to the next level. So you could have $var.strip()$, $var.format(#.##)# etc.

**Comment by jgm** [comment]
+++ richarddb [Apr 04 17 13:17 ]:
>   Would it make sense to support both? So something like $var$ will
>   output the variable as if it were text in the format; but $var.raw()$
>   wouldn't? The reason I'm suggesting the latter format, is that you
>   could extend it to other functions... I don't know if it's out of
>   bounds for pandoc, but I'm thinking there's some lightweight templating
>   functionality that would push it to the next level. So you could have
>   $var.strip()$, $var.format(#.##)# etc.

This has to be addressed before the template phase.  If we
have already parsed the content into a list of Inline
elements, there's no reliable way to turn this back into
the original string.

Hence #2139 addresses the issue at the level of the metadata
block.

**Comment by richarddb** [comment]
Thanks. Interesting discussion. I personally lean towards the "YAML should always be raw" interpretation:

- Because YAML is a _data_ serialisation format; not just a _string_ serialisation format.
- It's clearly in the header of the file, and processed "specially" as it is... not just plain markdown.
- It's always easier to "build up" a wrapped string compared to "stripping" or "decomposing" it.
- The ICML format is an example of where wrapping strings is not a commutative operation; the tags land up in the wrong place against the standard implementation.
- It means I can use different data types. For example, Booleans to selectively include/exclude blocks of content from the body.

The big downside seems to be an apparent loss of universality. But in my scheme, you would need template processing directives to "wrap" header values  appropriately for different target types, _which pandoc knows at runtime_.

Problem with my scheme: My guess is that defaulting to raw would break a lot of existing usages, if nothing else because that's the way it works now. So I would be tempted to vote for the "mark as raw in the header solution" for that reason; though it then gets more complicated if you want to use the variable as a boolean. Would we introduce typing?

But either way, it seems legitimate that there's valid use cases for both.

It does seem to me that either way this discussion starts moving pandoc in the direction of some generic templating features. I'm not entirely sure what that means, or if its "correct" per your vision?

Pro: Some users expect that functionality; and we refer to "templates".
Pro: It makes pandoc more capable and more useful.
Pro: A single "pan document processor" in one. Brilliant to author and multi-target.
Con: Adds complexity.
Con: It creates a "lossy" conversion from one format to the target—though that does happen elsewhere with other formats. Or can we say it's conceptually just a pre-processor, so no different to setting up my own toolchain with my own templating engine of choice?
Con: Arguments about templating syntax, feature set?

[I'm going to close this issue here, since it's already conceptually raised elsewhere.]
